File: 백준/Gold/1707. 이분 그래프/이분 그래프.cc
#include <cstring>
#include <vector>
#include <algorithm>
#include <iostream>
#include <queue>

#define ll long long

using namespace std;

int K, V, E;
vector<int> G[20001];
int Visited[20001];

void bfs(int start) {
    queue<int> q;

    Visited[start] = 1;
    q.push(start);

    while (!q.empty()) {
        auto curr = q.front();
        q.pop();

        for (auto &next: G[curr]) {
            if (Visited[next] == 0) {
                Visited[next] = Visited[curr] == 1 ? 2 : 1;
                q.push(next);
            }
        }
    }
}

bool is_valid() {
    for (int v = 1; v <= V; ++v) {
        for (auto &next: G[v]) {
            if (Visited[v] == Visited[next])
                return false;
        }
    }

    return true;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL), cout.tie(NULL);

    cin >> K;

    while (K--) {
        cin >> V >> E;

        int f, s;
        for (int e = 0; e < E; ++e) {
            cin >> f >> s;

            G[f].push_back(s);
            G[s].push_back(f);
        }

        for (int v = 1; v <= V; ++v) {
            if (Visited[v] == 0)
                bfs(v);
        }

        if (is_valid())
            cout << "YES" << '\n';
        else
            cout << "NO" << '\n';

        for (int v = 0; v <= V; ++v)
            G[v].clear();

        ::memset(Visited, false, sizeof(Visited));
    }

    return 0;
}

File: 백준/Gold/1707. 이분 그래프/이분 그래프.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

public class Main {

    private static int K, V, E;
    private static boolean[] VISITED = new boolean[20001];
    private static int[] MARK = new int[20001];
    private static List<Integer>[] G = new List[20001];

    private static boolean bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);

        VISITED[start] = true;
        MARK[start] = 1;

        while (!q.isEmpty()) {
            int curr = q.poll();
            for (Integer next : G[curr]) {
                if (!VISITED[next]) {
                    VISITED[next] = true;
                    MARK[next] = (MARK[curr] + 1) % 2;
                    q.add(next);
                } else {
                    if (MARK[curr] == MARK[next]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        K = Integer.parseInt(st.nextToken());

        while (K-- > 0) {
            st = new StringTokenizer(br.readLine());

            V = Integer.parseInt(st.nextToken());
            E = Integer.parseInt(st.nextToken());

            Arrays.fill(VISITED, false);
            Arrays.fill(MARK, 0);

            for (int i = 0; i <= V; i++) {
                G[i] = new ArrayList<>();
            }

            int u, v;
            for (int i = 0; i < E; i++) {
                st = new StringTokenizer(br.readLine());

                u = Integer.parseInt(st.nextToken());
                v = Integer.parseInt(st.nextToken());

                G[u].add(v);
                G[v].add(u);
            }

            boolean ret = true;
            for (int i = 1; i <= V; i++) {
                if (!VISITED[i]) {
                    if (!bfs(i)) {
                        ret = false;
                    }
                }
            }

            if (ret) {
                bw.write("YES\n");
            } else {
                bw.write("NO\n");
            }
        }

        br.close();
        bw.close();
    }
} 